excel_graph_parser/parser.py

```diff
@@ -0,0 +1,214 @@
+from io import BytesIO
+
+from openpyxl import load_workbook, Workbook
+from openpyxl.cell import Cell
+from viktor import File
+from viktor.external.spreadsheet import SpreadsheetCalculation
+
+import plotly.graph_objects as go
+
+ALLOWED_FIGURE_TYPES = ["lineChart", "scatterChart", "barChart", "pieChart"]
+
+
+class SpreadsheetParser:
+    def __init__(self, spreadsheet_calculation: SpreadsheetCalculation = None):
+        file = spreadsheet_calculation._file
+        if isinstance(file, File):
+            with file.open_binary() as r:
+                self.workbook = load_workbook(filename=r, data_only=True)
+        elif isinstance(file, BytesIO):
+            self.workbook = load_workbook(filename=file, data_only=True)
+        else:
+            raise NotImplementedError
+
+        self._spreadsheet_calculation = spreadsheet_calculation
+
+        # Gather charts by looping through sheets
+        self._charts_map = {}
+        untitled_index = 1
+        for sheet_name in self.workbook.sheetnames:
+            sheet = self.workbook[sheet_name]
+            for chart in sheet._charts:  # could be empty
+                if chart.title:
+                    chart_title = ''.join([title_element.t for title_element in chart.title.tx.rich.p[0].r])
+                else:
+                    chart_title = f"Untitled {untitled_index}"
+                    untitled_index += 1
+
+                self._charts_map[chart_title] = chart
+
+    def get_plotly_figure_by_title(self, title: str) -> go.Figure:

```

Comment:
I think we should re-purpose this to focus on the parsed data only, and not on the plotly figure:
```
def get_chart_data_by_title(...)
   ...
   return self._parse_chart_data(title, wb)
```

What do you think?

Reply:
The downside of this is that the VIKTOR app would require an extra line and separate import:
```
parser = SpreadsheetChartParser(...)
chart_data = parser.get_chart_data_by_title(title)
return PlotlyView(create_plotly_figure_from_chart_data(chart_data).to_json())
```

Reply:
I am not sure whether it makes sense for a developer to get the chart_data, whereas a go.Figure object is probably easier to manipulate if desired. Having to do `.to_json()` in the app is not very problematic I think. In the future we could make the PlotlyResult also accept go.Figure objects (or any object with a `to_json` handle)

Reply:
Let's do it!

Reply:
Updated